### PR TITLE
add shardingclient

### PR DESCRIPTION
**Pull request by duanhongyi**
Base on consistent hash, here's an example::

    from pymemcache.client import PooledClient, ShardingClient
    c1 = PooledClient(("10.1.15.12", 11211))
    c2 = PooledClient(("10.1.15.13", 11211))
    client = ShardingClient([c1, c2])
    client.set('key', {'a':'b', 'c':'d'})
    result = client.get('key')

### CODE REVIEW

**Comment by cgordon** [comment]
I'm on vacation right now, but will be back this coming weekend, and will review this then. Thank you for the commit!

**Comment by duanhongyi** [comment]
@cgordon

I will fix test case for py3.3

File: pymemcache/client.py [hunk 1]
```diff
@@ -113,11 +114,13 @@ def json_deserializer(key, value, flags):
 
 
 class MemcacheError(Exception):
+
     "Base exception class"
     pass
 
 
 class MemcacheClientError(MemcacheError):
+
```
Comment (cgordon) [inline pymemcache/client.py]: Please remove these extra empty lines in the Error classes (see the five below).

File: pymemcache/client.py [hunk 2]
```diff
@@ -999,6 +1009,256 @@ def __delitem__(self, key):
         self.delete(key, noreply=True)
 
 
+class ShardingClient(object):
+
+    def __init__(self, clients, black_list_timeout=30):
```
Comment (cgordon) [inline pymemcache/client.py]: Please document this method similarly to the constructors for PooledClient and Client.

File: pymemcache/client.py [hunk 3]
```diff
@@ -999,6 +1009,256 @@ def __delitem__(self, key):
         self.delete(key, noreply=True)
 
 
+class ShardingClient(object):
```
Comment (cgordon) [inline pymemcache/client.py]: The client.py file is getting rather large. Let's move ShardingClient to a new file called sharding_client.py. I'll do the same for PooledClient in a separate commit.

File: pymemcache/ring.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+from bisect import bisect
+
+if sys.version_info >= (2, 5):
+    import hashlib
+    md5_constructor = hashlib.md5
+else:
+    import md5
+    md5_constructor = md5.new
+
+class HashRing(object):
```
Comment (cgordon) [inline pymemcache/ring.py]: Did you write this code? If not, where did you get it? I just need to double check the licensing.

Comment (duanhongyi) [inline pymemcache/ring.py]: @cgordon 

Hash_ring source code from the end of 2012 no longer updated, it does not support py3k, so I update hash_ring.

Comment (duanhongyi) [inline pymemcache/ring.py]: @cgordon

From here:
https://pypi.python.org/packages/source/h/hash_ring/hash_ring-1.3.1.tar.gz

**Comment by duanhongyi** [comment]
@cgordon

merge && modified

**Comment by sontek** [comment]
@duanhongyi I started implementing an alternative to this in #59, want to take a look?  I also documented some of the reasoning behind it in #52

**Comment by duanhongyi** [comment]
close && see:

https://github.com/pinterest/pymemcache/pull/59

**Pull request closed without merging**